### DISCUSSION TITLE
Complete local only control research

### DISCUSSION THREAD

**Discussion by tacticallinux** [humbertogontijo/homeassistant-roborock#320] [General]
Regarding the issue #273 started by @jaydeethree I am experiencing a similar issue and hope to open a discussion about getting to full local only control of these vacuums. I have a Roborock Q5+ but I understand they all are similar.

In my use case I could care less about the Roborock/Xiaomi mobile apps or using them to monitor and control the device. My only interest is to be able to more or less fully control the vacuum via HA, at least the functionality needed to control it and schedule cleanings. I don't particularly care about mapping features and such although those would be cool.

It appears at least for my device that it cannot be controlled unless it can talk to mqtt-us-2.roborock.com and AWS although interestingly at the moment it does not have Internet access and shows it is Docked and stats are showing, don't know how accurate they are. I will do a manual spot clean here shortly using the physical buttons and see if stats get updated.

If I try to control the device via HA now HA reports it can't connect to the device with the following:

```
Logger: roborock.local_api
Source: runner.py:179
First occurred: 10:09:30 AM (330 occurrences)
Last logged: 10:42:00 AM

Failed connecting to 192.168.10.134: [Errno 111] Connect call failed ('192.168.10.134', 58867)
```

However if I jump on the HA box I can ping 192.168.10.134 and obviously the vacuum is connected to WiFi.

Anyone interested in working toward full local control of these devices? Love to hear your findings. If I can figure how to post a pic I will provide a simple diagram of my network.

**Comment by Lash-L** [comment]
It's an interesting problem I have not experimented with them, but we have a few commands that seem to do something with servers, I wonder if we could use them to our advantage? 

All these have to be lowercase to be sent.

```
    RoborockCommand.ENABLE_LOG_UPLOAD: CommandInfo(params=[9, 2]),
    RoborockCommand.SET_FDS_ENDPOINT: CommandInfo(params=["awsusor0.fds.api.xiaomi.com"]),
```

It's annoying as the devices obviously can work offline. I can get about 2 mqtt calls in before it does its network refresh. And I get dings confirming the vacuum received them and made the changes. 

It would be awesome if someone would be able to figure this out! As this is commonly brought up.

**Comment by jaydeethree** [comment]
To summarize the discussion so far from issue #273, the behavior I've seen on my Roborock S7 with firmware 02.16.12 is that if I block traffic to mqtt-us-2.roborock.com then the vacuum will frequently disconnect from my Wifi, reconnect, and send out a DHCPDISCOVER - it's effectively downing/upping its network interface. The frequency of this varies but I've seen it happen as quickly as every 30 seconds and as slowly as every 5 minutes. This happens regardless of whether I drop that traffic or block it (with a TCP RST). If I allow traffic to that domain on TCP port 8883 then the vacuum works normally, but of course that also means it's communicating with Roborock's servers.

It's worth noting that I also have an older Roborock E4 with ancient firmware (I have no idea which version, that info doesn't seem to be available in Home Assistant through the Xiaomi integration) and that one works fine when its Internet access is completely blocked. However it does something similar (frequently dropping off of my network and reconnecting to it) if I **drop** the traffic, but it works completely normally if I **block** the traffic (with a TCP RST). This is a well-known issue that's mentioned in Home Assistant's documentation for the Xiaomi Miio integration.

For now [Valetudo](https://valetudo.cloud/) seems to be the only known way to get these vacuums running 100% locally, but for most vacuum models you'll have to completely take the vacuum apart and short some pins together to put them into debug mode so you can flash it.

**Comment by humbertogontijo** [comment]
@jaydeethree can you check if you got any improvements in latest main?

**Reply by jaydeethree** [reply]
I'm very busy with work right now but I'll test this when I can - it should be sometime in the next few days.

**Reply by tacticallinux** [reply]
ditto, I got busy also with work but I will test soon. I looked over recent commits and didn't notice anything that may help our issue but I will test it out anyway.

**Reply by humbertogontijo** [reply]
I feel like the time using the vacuum after it is blocked from the internet increased. But it looks like it still rejects new connections after a while

**Reply by jaydeethree** [reply]
Okay, I just redownloaded the latest main, at commit `19f58af13b31aa1ffc61426ae24f435e64268f2f`. This error gets logged in Home Assistant during start-up:

```
Logger: homeassistant.setup
Source: setup.py:348
First occurred: 10:10:47 AM (1 occurrences)
Last logged: 10:10:47 AM

Unable to prepare setup for platform roborock.select: Platform not found (cannot import name 'DomainData' from 'custom_components.roborock.roborock_typing' (/config/custom_components/roborock/roborock_typing.py)).
```

But the integration still seems to work.

I then blocked all traffic from the vacuum to the Internet again and started running TCPDump - I can still see the vacuum frequently disconnecting/reconnecting to my network and sending out DHCPDISCOVERs. And as expected the vacuum becomes unavailable in Home Assistant when that happens. So from what I can see there is no change in behavior - it's still exactly the same as it was before.

**Reply by Reachushere** [reply]
Hi there,

Was there anything further to this?

My Q5 constantly becomes unavailable and I have to reset the wifi, readd it in to the integration etc... about every few days. I have to get a new token and new IP each time, its really irritating.  Is everyone else still experiencing this same issue, the last comment on this was a year and a half ago above and then it just seems to have stopped?

**Reply by daFreeMan** [reply]
In my very simplistic brain, would it be possible to add an entry to the HOSTS file on the HA server that redirects the vacuum's calls home to a call to a locally hosted server that knows how to respond properly?

I mean, HHCIB? 😁

**Comment by zastrixarundell** [comment]
This seems interesting https://github.com/Python-roborock/local_roborock_server. My Roborock is not supported by Valetudo
